mecanum_drive_controller/test/test_mecanum_drive_controller.cpp

```diff
@@ -0,0 +1,592 @@
+  ASSERT_EQ(
+    controller_->update_and_write_commands(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+
+  EXPECT_TRUE(std::isnan(controller_->reference_interfaces_[0]));
+
+  EXPECT_EQ(joint_command_values_[1], 0);
+  for (const auto & interface : controller_->reference_interfaces_)
+  {
+    EXPECT_TRUE(std::isnan(interface));
+  }
+  for (size_t i = 0; i < controller_->command_interfaces_.size(); ++i)
+  {
+    EXPECT_EQ(controller_->command_interfaces_[i].get_value(), 0.0);
+  }
+
+  std::shared_ptr<ControllerReferenceMsg> msg_2 = std::make_shared<ControllerReferenceMsg>();
+  msg_2->header.stamp = controller_->get_node()->now() - rclcpp::Duration::from_seconds(0.01);
+  msg_2->twist.linear.x = TEST_LINEAR_VELOCITY_X;
+  msg_2->twist.linear.y = TEST_LINEAR_VELOCITY_y;
+  msg_2->twist.linear.z = std::numeric_limits<double>::quiet_NaN();
+  msg_2->twist.angular.x = std::numeric_limits<double>::quiet_NaN();
+  msg_2->twist.angular.y = std::numeric_limits<double>::quiet_NaN();
+  msg_2->twist.angular.z = TEST_ANGULAR_VELOCITY_Z;
+  controller_->input_ref_.writeFromNonRT(msg_2);
+  const auto age_of_last_command_2 =
+    controller_->get_node()->now() - (*(controller_->input_ref_.readFromNonRT()))->header.stamp;
+
+  // age_of_last_command_2 < ref_timeout_
+  ASSERT_TRUE(age_of_last_command_2 <= controller_->ref_timeout_);
+  ASSERT_EQ((*(controller_->input_ref_.readFromRT()))->twist.linear.x, TEST_LINEAR_VELOCITY_X);
+  ASSERT_EQ(
+    controller_->update_reference_from_subscribers(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+  ASSERT_EQ(
+    controller_->update_and_write_commands(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+
+  EXPECT_NE(joint_command_values_[1], command_lin_x);
+  //  w0_vel = 1.0 / params_.kinematics.wheels_radius
+  // * (body_velocity_center_frame_.linear_x - body_velocity_center_frame_.linear_y
+  // - params_.kinematics.sum_of_robot_center_projection_on_X_Y_axis
+  // * body_velocity_center_frame_.angular_z);
+  //  joint_command_values_[1] = 1.0 / 0.5 * (1.5 - 0.0 - 1 * 0.0)
+  EXPECT_EQ(joint_command_values_[1], 3.0);
+  ASSERT_EQ((*(controller_->input_ref_.readFromRT()))->twist.linear.x, TEST_LINEAR_VELOCITY_X);
+  for (const auto & interface : controller_->reference_interfaces_)
+  {
+    EXPECT_TRUE(std::isnan(interface));
+  }
+
+  for (size_t i = 0; i < controller_->command_interfaces_.size(); ++i)
+  {
+    EXPECT_EQ(controller_->command_interfaces_[i].get_value(), 3.0);
+  }
+}
+
+// when in chainable mode and age_of_last_command < reference_timeout expect
+// reference_interfaces set by preceding controller and command_interfaces
+// are calculated to non-nan values and reference_interfaces are set to nan
+TEST_F(MecanumDriveControllerTest, test_update_logic_chainable_mode)
+{
```

Comment:
The following should be tested here:

1. activate controller.
1. Set some data to the references
1. Publish some (other) data to the reference message
1. Call `update` method
1. Expect references not being changed and commands being set from the values in the references and not in the message.

Reply:
i think point 5. is partially wrong denis as referenes will be set to nan after calling update and will not be same.

the rest is correct "commands being set from the values in the references and not in the message."

Reply:
I mean, the references are not updated from the input message.